portal/migrations/versions/69af4bd9ac9a_.py

```diff
@@ -0,0 +1,147 @@
+"""Validate and correct any QuestionnaireResponses needing attention
+
+Revision ID: 69af4bd9ac9a
+Revises: a1de3ab4d050
+Create Date: 2019-02-21 12:23:53.525121
+
+"""
+import copy
+from flask import current_app
+from flask_swagger import swagger
+import jsonschema
+import sys
+
+from portal.database import db
+from portal.system_uri import DECISION_SUPPORT_GROUP
+from portal.models.audit import Audit
+from portal.models.user import User
+from portal.models.questionnaire_response import QuestionnaireResponse
+
+# revision identifiers, used by Alembic.
+revision = '69af4bd9ac9a'
+down_revision = 'a1de3ab4d050'
+
+
+def validate_closure():
+    """redefine QuestionnaireResponse.validate_document for efficient loop"""
+    swag = swagger(current_app)
+
+    draft4_schema = {
+        '$schema': 'http://json-schema.org/draft-04/schema#',
+        'type': 'object',
+        'definitions': swag['definitions'],
+    }
+
+    validation_schema = 'QuestionnaireResponse'
+    # Copy desired schema (to validate against) to outermost dict
+    draft4_schema.update(swag['definitions'][validation_schema])
+
+    def validate(document, display_stack=False):
+        try:
+            jsonschema.validate(document, draft4_schema)
+            return True
+        except jsonschema.ValidationError as e:
+            if display_stack:
+                sys.stderr.write(str(e))
+            return False
+
+    return validate
+
+
+def lookup_system(label):
+    """fabricate a system from the identifier label"""
+    if label.startswith('WiserCare'):
+        return DECISION_SUPPORT_GROUP + '-wisercare'
+    raise ValueError("unknown identifer label {}".format(label))
+
+
+def correct_invalid_document(qnr_id, validate_fn, sys_id):
+    """Attempt to fix the invalid structure of a QNR.document
+
+    Many current QNRs are invalid as a user submitted w/o the `answer` list
+    as requires in FHIR DSTU2 QuestionnaireResponse.
+
+    Commits fix and audit row to db
+
+    """
+    qnr = QuestionnaireResponse.query.get(qnr_id)
+    document = copy.deepcopy(qnr.document)
+    answer_list = document['group']
+    if not answer_list:
+        raise ValueError("no answers under group for {}".format(qnr_id))
+
+    # each answer is to be a list of value_x_ objects
+    # nest in a list
+    for a in answer_list:
+        the_answer = a['answer']
+
+        # remove bogus 'system' at wrong level
```

Comment:
I love atomicity of each of these individual fixes. Would it be worthwhile to create an audit entry for each "fix" ran to make it obvious what's changed to an outsider?